// aws-cdk/test/api/aws-auth/sdk-logger.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const sdk_logger_1 = require("../../../lib/api/aws-auth/sdk-logger");
describe(sdk_logger_1.SdkToCliLogger, () => {
    const ioHost = {
        notify: jest.fn(),
        requestResponse: jest.fn(),
    };
    const logger = new sdk_logger_1.SdkToCliLogger(ioHost);
    beforeEach(() => {
        ioHost.notify.mockReset();
    });
    test.each(['trace', 'debug'])('%s method does not call notify', (method) => {
        logger[method]('SDK Logger test message');
        expect(ioHost.notify).not.toHaveBeenCalled();
    });
    test.each(['info', 'warn', 'error'])('%s method logs to notify', (method) => {
        logger[method]('SDK Logger test message');
        expect(ioHost.notify).toHaveBeenCalledWith(expect.objectContaining({
            level: 'trace',
            message: `[SDK ${method}] SDK Logger test message`,
        }));
    });
});
const SUCCESS_CALL = {
    clientName: 'S3Client',
    commandName: 'GetBucketLocationCommand',
    input: {
        Bucket: '.....',
        ExpectedBucketOwner: undefined,
    },
    output: { LocationConstraint: 'eu-central-1' },
    metadata: {
        httpStatusCode: 200,
        requestId: '....',
        extendedRequestId: '...',
        cfId: undefined,
        attempts: 2,
        totalRetryDelay: 30,
    },
};
const ERROR_CALL = {
    clientName: 'S3Client',
    commandName: 'GetBucketLocationCommand',
    input: {
        Bucket: '.....',
        ExpectedBucketOwner: undefined,
    },
    error: new Error('it failed'),
    metadata: {
        httpStatusCode: 200,
        attempts: 2,
        totalRetryDelay: 30,
    },
};
test('formatting a successful SDK call looks broadly reasonable', () => {
    const output = (0, sdk_logger_1.formatSdkLoggerContent)([SUCCESS_CALL]);
    expect(output).toMatchSnapshot();
});
test('formatting a failing SDK call looks broadly reasonable', () => {
    const output = (0, sdk_logger_1.formatSdkLoggerContent)([ERROR_CALL]);
    expect(output).toMatchSnapshot();
});
test('formatting a successful SDK call includes attempts and retries if greater than 1', () => {
    const output = (0, sdk_logger_1.formatSdkLoggerContent)([SUCCESS_CALL]);
    expect(output).toContain('2 attempts');
    expect(output).toContain('30ms');
});
test('formatting a failing SDK call includes attempts and retries if greater than 1', () => {
    const output = (0, sdk_logger_1.formatSdkLoggerContent)([ERROR_CALL]);
    expect(output).toContain('2 attempts');
    expect(output).toContain('30ms');
});
test('formatting a failing SDK call includes the error', () => {
    const output = (0, sdk_logger_1.formatSdkLoggerContent)([ERROR_CALL]);
    expect(output).toContain('it failed');
});
