// aws-cdk/test/context-providers/availability-zones.test.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const client_ec2_1 = require("@aws-sdk/client-ec2");
const api_1 = require("../../lib/api");
const availability_zones_1 = require("../../lib/context-providers/availability-zones");
const mock_sdk_1 = require("../util/mock-sdk");
const mockSDK = new (class extends mock_sdk_1.MockSdkProvider {
    forEnvironment() {
        return Promise.resolve({ sdk: new api_1.SDK(mock_sdk_1.FAKE_CREDENTIAL_CHAIN, mockSDK.defaultRegion, {}), didAssumeRole: false });
    }
})();
test('empty array as result when response has no AZs', async () => {
    // GIVEN
    mock_sdk_1.mockEC2Client.on(client_ec2_1.DescribeAvailabilityZonesCommand).resolves({
        AvailabilityZones: undefined,
    });
    // WHEN
    const azs = await new availability_zones_1.AZContextProviderPlugin(mockSDK).getValue({
        account: '1234',
        region: 'asdf',
    });
    // THEN
    expect(azs).toEqual([]);
});
test('returns AZs', async () => {
    // GIVEN
    mock_sdk_1.mockEC2Client.on(client_ec2_1.DescribeAvailabilityZonesCommand).resolves({
        AvailabilityZones: [{
                ZoneName: 'us-east-1a',
                State: 'available',
            }],
    });
    // WHEN
    const azs = await new availability_zones_1.AZContextProviderPlugin(mockSDK).getValue({
        account: '1234',
        region: 'asdf',
    });
    // THEN
    expect(azs).toEqual(['us-east-1a']);
});
